dcos/package.py

```diff
@@ -19,40 +19,43 @@ def uninstall(pkg, package_name, remove_all, app_id, cli, app):
     :type remove_all: boolean
     :param app_id: App ID of the app instance to uninstall
     :type app_id: str
-    :param init_client: The program to use to run the app
-    :type init_client: object
+    :param cli: Whether to remove the CLI, setting it to None will let
+                the program automatically decide (depending on the number
+                of app instances to remain after uninstallation)
+    :type remove_all: boolean
+    :param app: Whether to remove app
+    :type remove_all: boolean
     :rtype: None
     """
 
-    if cli is False and app is False:
-        cli = app = True
-
-    uninstalled = False
     installed = installed_packages(
-        pkg, app_id, package_name, cli_only=False)
-    installed_cli = next((True for installed_pkg in installed
-                          if installed_pkg.get("command")), False)
-    installed_app = next((True for installed_pkg in installed
-                          if installed_pkg.get("apps")), False)
-
-    if cli and installed_cli:
-        if subcommand.uninstall(package_name):
-            uninstalled = True
-
-    if app and installed_app:
-        if pkg.uninstall_app(package_name, remove_all, app_id):
-            uninstalled = True
+        pkg, None, package_name, cli_only=False)
+    installed_pkg = next(iter(installed), None)
+    installed_cli = installed_pkg.get("command") if installed_pkg else None
+    installed_app = installed_pkg.get("apps") if installed_pkg else []
 
-    if uninstalled:
-        return None
-    else:
+    if not installed_app and not installed_cli:
         msg = 'Package [{}]'.format(package_name)
         if app_id is not None:
             app_id = util.normalize_marathon_id_path(app_id)
             msg += " with id [{}]".format(app_id)
         msg += " is not installed"
         raise DCOSException(msg)
 
+    # If cli is set to None, try to be conservative and
+    # delete the CLI only when no apps would remain.
+    if cli is None:
+        cli = (remove_all or not app_id
+                or installed_app.count(app_id) == len(installed_app))
+
+    if app and installed_app:
+        if not pkg.uninstall_app(package_name, remove_all, app_id):
+            raise DCOSException("Couldn't uninstall package")
+
+    if cli and installed_cli:
+        if not subcommand.uninstall(package_name):
+            raise DCOSException("Couldn't uninstall subcommand")
+
```

Comment:
I don't really like the way the logic flows in this function. It's really hard to follow. Specifically, it's hard to understand the semantics around the values of `cli` and `app` (especially regarding the 3rd value of `None` that `cli` can now take on). It makes more sense to me to change the entire logic of this function to the following (where the comments match the code exactly):
```
# Passing `app == True` means that the user supplied an explicit `--app` flag on
# the command line. This forces an unconditional uninstall of the app associated
# with the supplied package (with different semantics depending on the values
# of `remove_all` and `app_id` as described in the docstring for this function).
if app and installed_app:
    if not pkg.uninstall_app(package_name, remove_all, app_id):
        raise DCOSException("Couldn't uninstall package")

# Passing `cli == True` means that the user supplied an explicit `--cli` flag on
# the command line. This forces an unconditional uninstall of the CLI associated
# with the supplied package.
if cli and installed_cli:
    if not subcommand.uninstall(package_name):
        raise DCOSException("Couldn't uninstall subcommand")

# Passing both `app == False` and `cli == False` means that the user didn't
# supply either `--app` or `--cli` on the command line. In this situation we
# uninstall the app associated with the supplied package (if it exists) as
# if the user had explicitly passed `--app` on the command line. However,
# we only uninstall the CLI associated with the package if the app being
# uninstalled is the last one remaining on the system. Otherwise, we leave the
# CLI in place so other instances of the app can continue to interact with it.
if installed_app:
    if not pkg.uninstall_app(package_name, remove_all, app_id):
        raise DCOSException("Couldn't uninstall package")

(Is this logic even correct? I don't understand why
`not app_id` is sufficient to imply all instances removed)
all_instances_removed = remove_all or \
                        not app_id or \
                        installed_app.count(app_id) == len(installed_app)

if installed_cli and all_instances_removed:
    if not subcommand.uninstall(package_name):
        raise DCOSException("Couldn't uninstall subcommand")
```
In general, overloading a boolean value to treat None as a 3rd option is very non-intuitive and should probably be avoided at all costs. Being explicit instead of clever can go a long way.

Reply:
> Is this logic even correct? I don't understand why `not app_id` is sufficient to imply all instances removed

Without the `--all` flag, not passing an app_id assumes there is a single app to uninstall, otherwise fails with `multiple apps installed, please set --all or --app-id`. So that condition seems correct to me. However, it can probably be simplified as : `remove_all or len(installed_app) <= 1` (meaning this is either a CLI only app, or a single app existed and was removed)